### PR TITLE
2022 01 18 issue 3969 Add `serialization_version` to `global_dlc_data`

### LINKED ISSUE #3969: Contract ids broken by changes to offer serialization (#3854)
Since we changed offer serialization in #3854 , all `contractId`s for numeric DLCs are broken in master. Likely we are going to have write migration code to update all `contractId`s at minimum, but likely we should update all stored DLCs that have the old serialization format.

>The temporary_contract_id MUST be the SHA256 hash of the offer_dlc message.

https://github.com/discreetlogcontracts/dlcspecs/blob/master/Protocol.md#requirements-1

I realized this when trying to get to the bottom of #3869

I get this error when trying to rescan on https://github.com/bitcoin-s/bitcoin-s/pull/3965/commits/ed9adb7771f6a294b19989b3be5038012502dff7 on my desktop. It seems to be a different error that the original issue. I'm not sure if it means this issue is fixed and this is a new unique issue, or of its related to new bug in #3965 

```
2022-01-08T17:12:55UTC INFO [DLCWallet$DLCWalletImpl] Found 1 matches from start=703891 to end=703990
2022-01-08T17:12:56UTC INFO [DLCWallet$DLCWalletImpl] Requesting 0 block(s)
2022-01-08T17:12:56UTC INFO [DLCWallet$DLCWalletImpl] Found 0 matches from start=703991 to end=704090
2022-01-08T17:12:56UTC INFO [DLCWallet$DLCWalletImpl] Found a match in block 704147
2022-01-08T17:12:56UTC INFO [DLCWallet$DLCWalletImpl] Requesting 1 block(s)
2022-01-08T17:12:57UTC INFO [DataMessageHandler] Received block message with hash 0000000000000000000d495f73083e0395ecdf432d8ffebff2d06789a8489722
2022-01-08T17:12:57UTC INFO [DLCWallet$DLCWalletImpl] Processing block=0000000000000000000d495f73083e0395ecdf432d8ffebff2d06789a8489722
2022-01-08T17:13:02UTC INFO [DLCWallet$DLCWalletImpl] Adding UTXO to wallet: de13c464b401ef130c7c67b2fa456b127f4461f0978957d63a12959c3cb6e98d:1 amt=33580 sats
2022-01-08T17:13:02UTC INFO [DLCWallet$DLCWalletImpl] Successfully inserted UTXO de13c464b401ef130c7c67b2fa456b127f4461f0978957d63a12959c3cb6e98d:1 into DB
2022-01-08T17:13:02UTC INFO [DLCWallet$DLCWalletImpl] Finished processing 1 received outputs, it took=187ms
2022-01-08T17:13:03UTC ERROR [DLCWallet$DLCWalletImpl] Error processing of block=0000000000000000000d495f73083e0395ecdf432d8ffebff2d06789a8489722.
java.lang.IllegalArgumentException: requirement failed: Offer and accept (without sigs) must refer to same event, offer.tempContractId=3924d10fa61467ffb2aabc9b9d718bb8efd04a3e93f338b8eacd8bc2877252a6 tempContractId=ca253ce6d58a0a3574ac957fe3c85b664166165c06f5ed869c322baf3344056d
        at scala.Predef$.require(Predef.scala:337)
        at org.bitcoins.core.protocol.dlc.build.DLCTxBuilder.<init>(DLCTxBuilder.scala:97)
        at org.bitcoins.dlc.wallet.internal.DLCDataManagement.$anonfun$builderFromDbData$8(DLCDataManagement.scala:434)
        at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:431)
        at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
        at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
        at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
2022-01-08T17:13:03UTC ERROR [DataMessageHandler] onBlockReceived Callback failed with error:
java.lang.IllegalArgumentException: requirement failed: Offer and accept (without sigs) must refer to same event, offer.tempContractId=3924d10fa61467ffb2aabc9b9d718bb8efd04a3e93f338b8eacd8bc2877252a6 tempContractId=ca253ce6d58a0a3574ac957fe3c85b664166165c06f5ed869c322baf3344056d
        at scala.Predef$.require(Predef.scala:337)
        at org.bitcoins.core.protocol.dlc.build.DLCTxBuilder.<init>(DLCTxBuilder.scala:97)
        at org.bitcoins.dlc.wallet.internal.DLCDataManagement.$anonfun$builderFromDbData$8(DLCDataManagement.scala:434)
        at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:431)
        at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
        at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
        at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

_Originally posted by @Christewart in https://github.com/bitcoin-s/bitcoin-s/issues/3869#issuecomment-1008068407_

### PR DESCRIPTION
**Pull request by Christewart**
fixes #3969

This PR adds a new column to `global_dlc_data` called `serialization_version`. 

When we fire up with dlc wallet with this change, this PR

1. Adds a new column to `global_dlc_data` called `serialization_version`.
2. Reads all rows inside of `global_dlc_data`
3. Queries for the `ContractInfo` associated with the row in `global_dlc_data`. 
4. Recomputes the `tempContractId` and `contractId` for DLCs
5. Updates them in the database

This is what the log looks like when the contractIds are migrated. This PR works because when I implemented #3854 I supported reading the old format (`Alpha`) and the new format (`Beta`), but we only support writing the new format (`Beta`). Now whenever we serialize something it will have an updated `tempContractId`/`contractId`. I then write the new `tempContractId`/`contractId` to the database. This migration also includes #4012 . All contractIds were broken in our current wallets because we did not include the funding output index in the hash computation.
 
The first few `WARN` logs are for #4001 , we cannot migrate those DLCs. See 4001 for why.

```
[info] 2022-01-26T18:46:07UTC WARN [DLCAppConfig] Corrupted dlcId=48c8fe360f09964920232452a315f7c621d9ebb602d19a22c2787ec6a7350958 state=Accepted, this is likely because of issue 4001 https://github.com/bitcoin-s/bitcoin-s/issues/4001 . This DLC will not have its contractId migrated to DLSerializationVersion.Beta
[info] 2022-01-26T18:46:07UTC WARN [DLCAppConfig] Corrupted dlcId=eafc75a2253898d60bc56900d3048aa774ffa49eb9bcfd0fe84d5740ca8c60c5 state=Accepted, this is likely because of issue 4001 https://github.com/bitcoin-s/bitcoin-s/issues/4001 . This DLC will not have its contractId migrated to DLSerializationVersion.Beta
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=3004f08a0fe2e2c928a78e33fd0f82ea431d6f576ac15004cd952a0802701cde old contractId=Some(b9abdebe933b1e27920cd9a111ef0c13aa9932d8489ca71fc70d912e0fcd27b6) new contractId=1bdb2ef6fd3b93cacf23de41d4e332330e670dc3104bb28e1fcda4b34f0ed402
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=9283c418e5dcd4bcffe4cc418de53b0d0f82cb685b75880ee3d86ea4951c18c5 old contractId=Some(4ee88b78d3331a9ef818c821b1549437edc9f0b55d15aca9d532925dc971a526) new contractId=4ee88b78d3331a9ef818c821b1549437edc9f0b55d15aca9d532925dc971a527
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=7f99783e13e2876e3ff481c684bdd45e51d350b9ebd7aae8bf7bfa328438485a old contractId=Some(0c13b4e890df982f1ef9700f0eeca222bd37fe3ccc4603c5771c60dcfaca1e5b) new contractId=bd7c2ff4936411910e9edc7780b6c3e642cdfa0770f968efabee2df98b96c5ab
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=770793bc0a18d6e03a2d739f1bab75d53f35cf0ac9c1037f6bcf639758677053 old contractId=Some(f7d79356749d1817a16d9fb85024ca4e42e07e6af166a075c1ffe12ff71abdd2) new contractId=f7d79356749d1817a16d9fb85024ca4e42e07e6af166a075c1ffe12ff71abdd3
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=3ab3d9dab52d4042307ac300d492ffe93a7549554fa9c5cec6266498512e7752 old contractId=Some(17bc62d208b60046e7919d46058da30edb9792b7995d89b3e8c6c7f2db1a6de9) new contractId=5cdbe3d538b3b878e39a5b1b54cde50719099fd2ce60c58b6777dd26f7d7fb2a
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=812252cf184c787bb578064a2444cd35d3fe0fb05e4cc970e5d96abc7f5e610e old contractId=Some(d1191b88427a60a92e5f9c5afab132b7d67541ef13a009e41710491e7f110c7e) new contractId=9505d226ab29991a2fa5b9acb411980b36f17b430c570afd33f8b777d3c5e5ef
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=eb514aeb62a7c95495372592c99339e1de3f4e2289cb30fc11e274301896e99c old contractId=Some(fd6443ffde9783cd2f954a9776919e27533bacdec8ef746347c6af1f911a0c3d) new contractId=4e2595ec6a01cc737fddb8d09268521ccbf466db5c031ec38d2a152b3bfa0fb9
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=1a20fa18a46d6c1e1b4dd303fc0a9c95adb4b23674066a1e73291d933474a296 old contractId=Some(e28cdd8a5d244f7ba2ed57fe93ca7395ee6e32748250bd224c8add3409989918) new contractId=7747524b66e3a25cd939773beeb70b95807bdaa75bf3b5bcb2752268e361140d
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=b66003a33c5c84e40ffcf7cfe6d6082c3622493e4feaafcf2a2fbe68dd6ea1c3 old contractId=Some(2c43fb0aaa07383f5ad40fade2c1876f9758af822a799ca432d1c47d14ce69f9) new contractId=38909056e88029bff7a82e301a68046d661d8479bac738e97cb73ed72d0d0a2d
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=2ba1a3a5a3c9a9218e530db82529cabd0afaa56dd19a0173112556b75d9054c9 old contractId=Some(888b0c0445f26ad8322934337d9571b550831b324f91a84ea85f96af6aa0c67b) new contractId=db127290256dad3ae869c21387990a152a4f2fe447982a62c2200f6a9f616469
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=9dfcb710dbe63d7d2ccbf52354a92e0c041beaddf4d04ca1921eb33055dfdeea old contractId=Some(b641e3bc1eaa25b1e2f3199f118ec60ed8d50340c7cf3a6e1b20ed6252b25da1) new contractId=b641e3bc1eaa25b1e2f3199f118ec60ed8d50340c7cf3a6e1b20ed6252b25da3
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=2543d091c657ca167e0da2ba455006217dd3e965520679c3dc5ad2d190ba3c76 old contractId=Some(43e666d806e0fd4608110096ba5c4c144e4855d93ee37a67b3006992fdb2e3d6) new contractId=43e666d806e0fd4608110096ba5c4c144e4855d93ee37a67b3006992fdb2e3d7
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=201f6fc4922520c54c11aa315c4beecc8eaad867696cb36d9899adda4b2d82ff old contractId=Some(0a07a525115ea99d058464d87451e88db83465e451005dfe15acaaa475c7872f) new contractId=6495bf5314ab9e7a8d1b9fe902eda263b7b4198aa8c40da052eeeebd2781bda3
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=bf5c7f772c14b147a0d8701e0f5cb056c856305b9eec3c3b00ed5bc54eb7757e old contractId=Some(8c755590a3d281553960efe2f6691b6a8ed02354fc265ba8364597b830a6e3a0) new contractId=f16c89e1a25b03c920a6f72fc071e9b2f5283f2c0f1d51b48c61d553274c2a12
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=ffdf77047ccb5ed938fcbcee79e1da6d6a294bc593d4dfc12f8c94077dcfdcdd old contractId=Some(f502202a83c47f2f139e2c564dee942780bbedb893759103cf37b140480bac2b) new contractId=88f6cdf59de1dfce85e87e93fa6d717be7bab303c67d6eff2898d7f9e09e1070
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=2d873c17eb7dea881dd1f5620ad945725bce9e063ee4ec05c9047b4a9036599c old contractId=Some(5536279d992f12d6811aa9d357b308cf078255fb10cdea203672eacb4293f7e5) new contractId=5536279d992f12d6811aa9d357b308cf078255fb10cdea203672eacb4293f7e4
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=a37924fb9da76ed62a0788555a75b232ca1c12801ecf4179685d1545b7d73c1c old contractId=Some(c6cf9ea699afb840dd48e2fbdc9420a1b5b82ca342fbb06432ab76e2edb914c8) new contractId=c6cf9ea699afb840dd48e2fbdc9420a1b5b82ca342fbb06432ab76e2edb914c9
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=6718226c9142a0f4e3db8e9a9194e3aaefa0e61913591f8b21a738eeab7622a5 old contractId=Some(1d9e7a51ab60a9574730caa301246cf09ae45c2cd72bc463d616f1efb79c860f) new contractId=6abcf1872f40022bba37ef951c7b2afda5138ecdb669b427c185e28546929bfe
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=7331e9d8adbb479bf604580ea359bcd9022d5ea9fde94115355d14dc55aeb9a5 old contractId=Some(9deff3fa22764d6e31b1ed98e538cede6325fb285093542b8fff8a195101ba1c) new contractId=86ecddfc5def6e1465ec309528e2b355290c48d76cbf9b8baf4b2c6a32a4e0e4
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=07d162ede4e50159167956af6f168da93676edea7ff179e603bd100a75596b10 old contractId=Some(2d52d31c7176f19743d53e20281e1fca7aa3d7a07b7098d70cc0aa78f70a1d75) new contractId=607cb0e7a1b92893abbb5a032fe673eb17c2e8865f1e7e64d3912e3ef2af14b9
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=6bd72e08ee1f39def7901df78a34ae6c841a776cfeb5dc39637d4ec21d296d45 old contractId=Some(4e71aa624a2b03292eb4e9c3a52f28c029d40e2b60151c53170421184c328393) new contractId=4e71aa624a2b03292eb4e9c3a52f28c029d40e2b60151c53170421184c328392
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=fc1e685d00e6ddd2d9a5e6b8c1745ef14276457e23b4c9b270a05c4726baf20e old contractId=Some(90b8221b09a9d341ef8161f2d3fe4c813778bf0774175a2dc327975a959fc02d) new contractId=caac25118d229cce67ba028b96ae951fb2b80a0e9f1eb93cbf163fb4ef381c67
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=c1b8d0f61e57b5059db72e4bdb694f0b87c1617e0f6edb2627213806ef91b2f5 old contractId=Some(721093f57e4518bd292df18f9e62ddfe1a5a6a569d89dce973e147ac31976a3f) new contractId=5649fb1468c4a449ba8030e024653a8f60b474f964e1ae3e4ed3e0998ea59a1e
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=b5a9092d9c033dd4f7e05a67785455b8d126ca244e32be9f6d9712d71d761b04 old contractId=Some(056e37b986b833b0ed645537de303f88bd9b9a900bebe3dec11bb4748e22f499) new contractId=c2633fe68979803fbeb8dd352b4e1a529747f23b072601125c552b459b595491
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=2f1224d3ab1f2456d020e7777009153e2ee5db2d127b94774580e428f5a899b9 old contractId=Some(a995aeca167f1c28c2b936e5bb4d87cba061dd9147217501c3523e377fd23439) new contractId=a995aeca167f1c28c2b936e5bb4d87cba061dd9147217501c3523e377fd23438
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=543249d7c34014a774810d3633fb7d19f339fbcd4e88ceccad7018fa7b611706 old contractId=Some(fc4807e44939b83576b50e2ae6e12ba84bdd4ff398addb07f91a3dc045c62701) new contractId=06507199a88b64dddaa51897f152cef34e6d1988a6efd96344cb35bda2b70532
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=ecec5ef9f0fc7357865316dd6dbdcd9b28bbbddb6a8132d984dcbfb1c39770d5 old contractId=Some(41ea8729fcdda88750c12b8736b4c71feaf006a80f2e6b07a35a23d7d2e5a0d6) new contractId=41ea8729fcdda88750c12b8736b4c71feaf006a80f2e6b07a35a23d7d2e5a0d4
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=df11a4fa70ce38f2309ce80727daea9bc64e33bc9f6477c07718728a1e3576c4 old contractId=Some(92c92ceb848c6cee6b01f97770340566bec99dd132a3df2d32cd6a22d628680d) new contractId=92c92ceb848c6cee6b01f97770340566bec99dd132a3df2d32cd6a22d628680c
[info] 2022-01-26T18:46:08UTC INFO [DLCAppConfig] Updating contractId for dlcId=9f3c815184530e0f1f1fa6da39d4a7ef5c7501b5624cfc0ab85bbe82ba5ae40e old contractId=Some(4c81d2b44e22e13581550acfadbfbd4c6d73f0b45650ffa45afca202603999fc) new contractId=4c81d2b44e22e13581550acfadbfbd4c6d73f0b45650ffa45afca202603999fd


```

I've migrated 4 wallets of mine on  a4762b9328d09a5785eaa8d1533166a23e7b194f

### CODE REVIEW

File: dlc-wallet/src/main/resources/sqlite/dlc/migration/V5__serialization_version_global_dlc_data.sql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE "global_dlc_data"
+    ADD COLUMN "serialization_version" VARCHAR(254);
```
Comment (benthecarman) [inline dlc-wallet/src/main/resources/sqlite/dlc/migration/V5__serialization_version_global_dlc_data.sql]: I think it'd make sense to have this be `NOT NULL` with a default of Alpha and then update the correct ones accordingly. That way the schema enforces that a `serialization_version` must be set

**Comment by benthecarman** [comment]
> This PR does not currently upgrade the broken contractIds. I'm undecided whether to add that to this PR or follow up with another PR. We will see how long it takes to get this merged.


imo this should happen the same pr, otherwise the migration code can get very hairy

**Comment by Christewart** [comment]
I successfully negotiated a DLC end to end on cec7b66b42598a6ea9e06211ba543e477c1bac12 between my mac and desktop.

https://mempool.space/tx/d17c8147ecf30f11f745bf20d06b76786493e0d36dc0bfd26c62bccfc291d15b

**Comment by Christewart** [comment]
Run by @benthecarman @user411 @rorp and everyone seemed to upgrade their wallet correctly :tada: . Merging

**Pull request merged**